### ISSUE TITLE
high-level feedback from Google doc as of morning Mon Dec 9 2024

### ISSUE THREAD

**Issue by jowens** [b0nes164/Decoupled-Fallback-Paper#1]
### Ingredients of an introduction

There are four:
1. What is the exact problem that is being solved?
2. Why is this problem important?
3. Why is this problem hard / why aren't the current solutions to this problem sufficient? (In our case, that's going to be an assertion that current techniques don't work without fpg, but we will certainly not explain it in detail at this point; that's going to be a whole section later in the paper.)
4. What are the precise contributions of the work?

### Separate the description into "design" and "implementation"

Most of this paper is going to be explaining the techniques. That is a big chunk of text. The way we typically factor that is a section on the design of the method/s and a separate section on how it is implemented. Roughly, "design" can be explained to someone who has a reasonable knowledge of data-parallel primitives, but nothing about CUDA/WebGPU.

If we adopt that idea, we can factor the overall description of the work in two ways:

1. Design section, implementation section, which is
    1. Overall architecture
    5. Design of all the parts in our system, one by one.
    6. Implementation of all the parts in our system, one by one.
2. Per-part design, implementation, which is
    1. Overall architecture
    2. Design of one part in our system, then its implementation
    3. Design of a second part in our system, then its implementation
    7. ...

We don't have to decide this right away. The former organization is generally how it's done.

### High-level design comes first

The current draft makes a lot of assumptions about the architecture of the scan and drops (way too) quickly into how it's implemented. We want to instead start with [goals / non-goals / constraints](https://graphics.stanford.edu/~kayvonf/notes/systemspaper/) of the overall system. Then we say given these goals / non-goals / constraints, we adopt the following high-level architecture (e.g., one that is built around a root / follows a stream-scan-like approach). If this is an existing architecture and it's the right architecture for our goals / non-goals / constrains, we can just say that. We don't have to reexplain existing work. But any high-level architectural differences are ones that we have to motivate and explain, carefully.

THEN, only after the high-level architecture is established, we can start talking about the lower-level details of the design and implementation. For instance, we have the additional constraint that our implementation has to be subgroup-size-agnostic. Why is the current approach insufficient to satisfy that constraint?

cc: @raphlinus

**Comment by raphlinus** [comment]
Agree with what John wrote above, and have some other thoughts. I've marked up my copy of the paper with finer-grained edits, but perhaps those will work best as a PR against the repo.

The main story we need to tell in the introduction etc is: why can't you implement decoupled lookback portably? Boiled down to the core, that's three things:

* For any monoid larger than 30 bits, you need a device-scope memory barrier, for correct operation. That barrier does not exist on Metal pre-3.2, and does not exist in WebGPU.
* Even on GPUs and APIs which do support the barrier, it's a performance slowdown (we'll need empirical data later in the paper to support this).
* Even if you had both of the above (or your monoid fits in 30 bits), running on a device without FPG means you might run very slow, or deadlock. (we don't know even at runtime whether a device has FPG, because, even though a query that has been proposed by Sorensen et all, it hasn't been adopted).

(re the monoid size argument, cite that part of the decoupled lookback paper; it mentions being able to use 64 bit atomics)

There's a separate story about subgroups, which is important to the overall portability narrative, but I think that should be cleanly factored out. Importantly, it doesn't sink the algorithm, just holds back performance. (and an annoying caveat is that you can't rely on subgroups being present in WebGPU). But basically it's orthogonal 

So basically the current state of play is that anyone wanting to implement scan on WebGPU has to rely on reduce-then-scan, as decoupled lookback is not viable. The main problem with that is that it's slower. Secondarily, you have to do multiple dispatches, and the number of dispatches depends on the size of the problem (if you're going to avoid a serial step).

So then the next part of the argument is that we can systematically solve those issues. I would focus the discussion on the "status counter array" implementation of the "pass message from one workgroup to another" primitive as an alternative to the standard acquire/release technique. Then we can say that we build on that primitive; it's not a new algorithm in the way that decoupled lookback was, and that's a *good* thing. But in any case it's always a good idea to be super clear on what's the new part and what we're building on.

Fallback is another part of the story, and in some ways orthogonal, but for some reason there's overlap between devices with poor scheduler fairness and ones that don't have the barrier. Then, later in the paper, we'll want to break that down into two parts: doing fallback at all moves the algorithm from "may deadlock" to "will produce a correct result assuming your GPU works". But you *also* need the fallback to be nice and parallel to avoid slowdown.

I have some more thoughts, but wanted to send what I had.
